project_1/include/utils.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <cstdint>
+#include <iostream>
+#include <vector>
+
+enum event_type: std::uint8_t {
+    SEND_DATA,
+    ACKNOWLEDGEMENT,
+};
+
+// Type for packet
+struct PacketHeader {
+private:
+    std::uint64_t sender_id;
+    std::uint64_t receiver_id;
+    std::uint64_t sending_time;
+    std::uint32_t packet_index;
+    std::uint32_t estimated_delivery_time;
+public:
+    PacketHeader() = default;
+    PacketHeader(std::uint64_t, std::uint64_t, std::uint64_t, std::uint32_t, std::uint32_t);
+
+    std::uint32_t GetEstimatedDeliveryTime() const;
+    std::uint64_t GetSendingTime() const;
+    std::uint32_t GetPacketIndex() const;
+    std::uint64_t GetSenderID() const;
+
+    bool operator<(const PacketHeader& other) const;
+};
+
+// Type for congestion window
+struct CWND {
+private:
+    std::uint32_t size_in_packets = 8;
+
+    // Slow-start threshold in packets
+    std::uint32_t ssthresh = 1024;
+public:
+    CWND() = default;
+
+    std::uint32_t GetSize() const;
+    void IncreaseSize();
+};
+
+// Type for server receiver, contains only ID
+struct ServerBase {
+protected:
+    std::uint64_t id;
+public:
+    ServerBase(std::uint32_t);
+    ServerBase() = default;
+    ~ServerBase() = default;
```

Comment:
```suggestion
    virtual ~ServerBase() = default;
```

Reply:
Guideline #4: A base class destructor should be either public and virtual, or protected and nonvirtual.
from http://www.gotw.ca/publications/mill18.htm